src/drivers/riscv,cpu.c

```diff
@@ -0,0 +1,540 @@
+/* Copyright 2018 SiFive, Inc */
+/* SPDX-License-Identifier: Apache-2.0 */
+
+#include <mee/io.h>
+#include <mee/shutdown.h>
+#include <mee/machine.h>
+
+struct mee_cpu *__mee_driver_cpu_get(int hartid)
+{
+    if (hartid < __MEE_DT_MAX_HARTS) {
+        return &(__mee_cpu_table[hartid].cpu);
+    }
+    return (struct mee_cpu *)NULL;
+}
+
+unsigned long __mee_myhart_id (void)
+{
+    unsigned long myhart;
+    asm volatile ("csrr %0, mhartid" : "=r"(myhart));
+    return myhart;
+}
+
+void __mee_zero_memory (unsigned char *base, unsigned int size)
+{
+    volatile unsigned char *ptr;
+    for (ptr = base; ptr < (base + size); ptr++){
+        *ptr = 0;
+    }
+}
+
+void __mee_interrupt_global_enable (void) {
+    unsigned long m;
+    asm volatile ("csrrs %0, mstatus, %1" : "=r"(m) : "r"(MEE_MIE_INTERRUPT));
+}
+
+void __mee_interrupt_global_disable (void) {
+    unsigned long m;
+    asm volatile ("csrrc %0, mstatus, %1" : "=r"(m) : "r"(MEE_MIE_INTERRUPT));
+}
+
+void __mee_interrupt_software_enable (void) {
+    unsigned long m;
+    asm volatile ("csrrs %0, mie, %1" : "=r"(m) : "r"(MEE_LOCAL_INTERRUPT_SW));
+}
+
+void __mee_interrupt_software_disable (void) {
+    unsigned long m;
+    asm volatile ("csrrc %0, mie, %1" : "=r"(m) : "r"(MEE_LOCAL_INTERRUPT_SW));
+}
+
+
+void __mee_interrupt_timer_enable (void) {
+    unsigned long m;
+    asm volatile ("csrrs %0, mie, %1" : "=r"(m) : "r"(MEE_LOCAL_INTERRUPT_TMR));
+}
+
+void __mee_interrupt_timer_disable (void) {
+    unsigned long m;
+    asm volatile ("csrrc %0, mie, %1" : "=r"(m) : "r"(MEE_LOCAL_INTERRUPT_TMR));
+}
+
+void __mee_default_exception_handler (void) {
+    mee_shutdown(100);
+}
+
+void __mee_default_interrupt_handler (int id, void *priv) {
+    mee_shutdown(200);
+}
+
+void __mee_default_sw_handler (int id, void *priv) {
+    unsigned long mcause;
+    struct __mee_driver_riscv_cpu_intc *intc;
+    struct __mee_driver_cpu *cpu = &__mee_cpu_table[__mee_myhart_id()];
+
+    asm volatile ("csrr %0, mcause" : "=r"(mcause));
+    if ( cpu ) {
+        intc = (struct __mee_driver_riscv_cpu_intc *)cpu->interrupt_controller;
+        intc->mee_exception_table[mcause & MEE_MCAUSE_CAUSE]();
+    }
+}
+
+void __mee_default_timer_handler (int id, void *priv) {
+    struct mee_cpu *cpu = __mee_driver_cpu_get(__mee_myhart_id());
+    unsigned long long time = __mee_driver_cpu_mtime_get(cpu);
+
+    /* Set a 10 cycle timer */
+    __mee_driver_cpu_mtime_set(cpu, time + 10);
+}
+
+void __mee_exception_handler(void) __attribute__((interrupt, aligned(128)));
+void __mee_exception_handler (void) {
+    int id;
+    void *priv;
+    unsigned long mcause, mepc, mtval;
+    struct __mee_driver_riscv_cpu_intc *intc;
+    struct __mee_driver_cpu *cpu = &__mee_cpu_table[__mee_myhart_id()];
+
+    asm volatile ("csrr %0, mcause" : "=r"(mcause) );
+    asm volatile ("csrr %0, mepc" : "=r"(mepc));
```

Comment:
We need to advance MEPC by four bytes so that execution resumes after the faulting instruction if the exception handler returns.

Reply:
Ooh, hold on. The instruction could be compressed, in which case we'd need to advance two bytes.